server/src/main/java/org/elasticsearch/rest/MethodHandlers.java

```diff
@@ -31,36 +32,40 @@
 final class MethodHandlers {
 
     private final String path;
-    private final Map<RestRequest.Method, RestHandler> methodHandlers;
+    //TODO maybe we should aim for having a type for version instead of string/byte
+    private final Map<RestRequest.Method, Map<String,RestHandler>> methodHandlers;
 
-    MethodHandlers(String path, RestHandler handler, RestRequest.Method... methods) {
+    MethodHandlers(String path, RestHandler handler, String version, RestRequest.Method... methods) {
         this.path = path;
         this.methodHandlers = new HashMap<>(methods.length);
         for (RestRequest.Method method : methods) {
-            methodHandlers.put(method, handler);
+            methodHandlers.putIfAbsent(method, new HashMap<>());
+            methodHandlers.get(method).put(version, handler);
```

Comment:
nit: this can be simplified to ` methodHandlers.computeIfAbsent(method, k -> new HashMap<>()).put(version, handler);` (mostly a cosmetic change though...)

Reply:
good idea!